File: src/algorithms/python/actorcritic.py
# Actor-Critic on a gridworld.
#
# Actor-Critic combines the two families: a *critic* learns a state-value table
# V[s] (like value methods), and an *actor* learns a softmax policy theta[s,a]
# (like REINFORCE). At every step the critic supplies a one-step TD error that
# acts as a low-variance advantage signal for the actor:
#
#     delta = r + gamma * V[s'] - V[s]            (TD error / advantage)
#     V[s]      <- V[s]      + lr_critic * delta
#     theta[s,.] <- theta[s,.] + lr_actor * delta * grad log pi(a|s)
#
# Bootstrapping with the critic means the actor is updated every step from a
# learned baseline rather than waiting for a full Monte-Carlo return, so it has
# far lower variance than REINFORCE. The heatmap shows the critic's V[s].
#
# Environment cell codes: 0 empty . 1 wall . 2 goal (+1) . 3 trap (-1) . 4 start

import numpy as np

# up, right, down, left
ACTIONS = [(-1, 0), (0, 1), (1, 0), (0, -1)]


def run(X, y=None, episodes=70, lr_actor=0.2, lr_critic=0.2, gamma=0.95, seed=0):
    rng = np.random.default_rng(seed)
    grid = np.asarray(X).astype(int)
    rows, cols = grid.shape
    nS, nA = rows * cols, 4

    starts = np.argwhere(grid == 4)
    sr, sc = (int(starts[0][0]), int(starts[0][1])) if len(starts) else (0, 0)
    start = sr * cols + sc

    def step(s, a):
        r, c = divmod(s, cols)
        dr, dc = ACTIONS[a]
        nr, nc = r + dr, c + dc
        if nr < 0 or nr >= rows or nc < 0 or nc >= cols or grid[nr, nc] == 1:
            nr, nc = r, c
        ns = nr * cols + nc
        cell = grid[nr, nc]
        if cell == 2:
            return ns, 1.0, True
        if cell == 3:
            return ns, -1.0, True
        return ns, -0.025, False

    def softmax(z):
        z = z - z.max()
        e = np.exp(z)
        return e / e.sum()

    theta = np.zeros((nS, nA))   # actor preferences
    V = np.zeros(nS)             # critic values
    max_steps = 4 * nS

    def snapshot():
        greedy = theta.argmax(axis=1)
        vals, pol = [], []
        for s in range(nS):
            r, c = divmod(s, cols)
            if grid[r, c] in (1, 2, 3):
                vals.append(0.0)
                pol.append(-1)
            else:
                vals.append(float(V[s]))      # critic's state value
                pol.append(int(greedy[s]))
        return vals, pol

    yield {
        "type": "rl:init",
        "step": 0,
        "label": "Actor-Critic",
        "rows": rows,
        "cols": cols,
        "grid": grid.flatten().tolist(),
        "nActions": nA,
        "explanation": (
            f"Actor-Critic on a {rows}\u00d7{cols} gridworld. A critic learns state values "
            f"V[s] while an actor learns a softmax policy; the critic's TD error tells the "
            f"actor which actions beat expectation. Cells are shaded by V[s]."
        ),
        "math": r"\delta=r+\gamma V(s')-V(s)",
    }

    for ep in range(int(episodes)):
        s = start
        path = [s]
        total = 0.0
        for _ in range(max_steps):
            p = softmax(theta[s])
            a = int(rng.choice(nA, p=p))
            ns, r, done = step(s, a)

            target = r if done else r + gamma * V[ns]
            delta = target - V[s]                 # TD error = advantage
            V[s] += lr_critic * delta

            grad = -p
            grad[a] += 1.0                         # d log pi / d theta[s]
            theta[s] += lr_actor * delta * grad

            total += r
            s = ns
            path.append(s)
            if done:
                break

        vals, pol = snapshot()
        yield {
            "type": "rl:episode",
            "step": ep + 1,
            "iteration": ep,
            "label": "Actor-Critic",
            "episode": ep,
            "values": vals,
            "policy": pol,
            "reward": float(total),
            "steps": len(path) - 1,
            "path": [int(p) for p in path],
            "explanation": (
                f"Episode {ep}: reward {total:+.2f} in {len(path) - 1} steps. The critic's "
                f"values (cell shading) and the actor's arrows improve together each step."
            ),
            "math": r"\theta_s\mathrel{+}=\alpha_a\,\delta\,(\mathbb{1}[a]-\pi);\;\; V(s)\mathrel{+}=\alpha_c\,\delta",
        }

    vals, pol = snapshot()
    yield {
        "type": "rl:converged",
        "step": int(episodes) + 1,
        "label": "Actor-Critic",
        "values": vals,
        "policy": pol,
        "episodes": int(episodes),
        "reason": f"Ran {int(episodes)} episodes; actor and critic converged together.",
        "explanation": (
            "Done. By bootstrapping off the critic, the actor learns from a low-variance "
            "advantage every step \u2014 typically faster and steadier than Monte-Carlo "
            "REINFORCE."
        ),
        "math": r"\nabla_\theta J=\mathbb{E}\big[\,\delta\,\nabla_\theta\log\pi_\theta(a\mid s)\,\big]",
    }

File: src/algorithms/actorcritic.ts
import type { AlgorithmMeta } from '@/types/algorithm';

export const actorcriticMeta: AlgorithmMeta = {
  id: 'actorcritic',
  family: 'rl',
  name: 'Actor-Critic',
  shortDescription: 'Combines a policy (actor) with a learned value baseline (critic) for low-variance updates.',
  longDescription:
    "Actor-Critic fuses the two families. A critic learns a state-value table V[s] (like value methods), while an actor learns a softmax policy θ[s,a] (like REINFORCE). At every step the critic supplies a one-step temporal-difference error δ = r + γ·V[s′] − V[s] that doubles as a low-variance advantage signal: the critic is nudged toward δ and the actor's preferences are pushed by δ·∇log π(a|s). Because the actor is updated every step from a learned baseline — rather than waiting for a full Monte-Carlo return — it has far lower variance than REINFORCE and usually learns faster and more steadily. The heatmap shows the critic's V[s]; the arrows show the actor's greedy policy.",
  category: 'reinforcement',
  task: 'reinforcement',
  pythonFilename: 'actorcritic.py',
  sklearnSnippet: `# Modern actor-critic (A2C) via Stable-Baselines3.
import gymnasium as gym
from stable_baselines3 import A2C

env = gym.make("CartPole-v1")
model = A2C(
    "MlpPolicy", env,
    learning_rate=7e-4,
    gamma=0.99,
    n_steps=5,          # TD bootstrap horizon
)
model.learn(total_timesteps=100_000)`,
  hyperparams: [
    { id: 'episodes', label: 'Episodes', codeKey: 'episodes=', type: 'int', min: 30, max: 150, step: 10, default: 70 },
    { id: 'lr_actor', label: 'Actor LR', codeKey: 'lr_actor=', type: 'float', min: 0.01, max: 0.5, step: 0.01, default: 0.2 },
    { id: 'lr_critic', label: 'Critic LR', codeKey: 'lr_critic=', type: 'float', min: 0.01, max: 0.5, step: 0.01, default: 0.2 },
    { id: 'gamma', label: 'Discount (γ)', codeKey: 'gamma=', type: 'float', min: 0.5, max: 0.99, step: 0.01, default: 0.95 },
  ],
  timeComplexity: 'O(episodes · steps)',
  spaceComplexity: 'O(states · actions)',
  pros: ['Low-variance updates via the critic baseline', 'Learns online, every step', 'Bridges value and policy methods'],
  cons: ['Two interacting learners to tune', 'Critic bias can mislead the actor', 'Less stable than pure value methods if mistuned'],
  compatibleTasks: ['reinforcement'],
  references: [
    { kind: 'wiki', label: 'Wikipedia: Actor-critic algorithm', url: 'https://en.wikipedia.org/wiki/Actor-critic_algorithm' },
    { kind: 'paper', label: 'Mnih et al. (2016): Asynchronous Methods (A3C)', url: 'https://arxiv.org/abs/1602.01783' },
    { kind: 'article', label: 'Sutton & Barto: Reinforcement Learning (Ch. 13.5)', url: 'http://incompleteideas.net/book/the-book-2nd.html' },
  ],
};
